src/coins.cpp

```diff
@@ -265,22 +283,84 @@ bool CCoinsViewCache::BatchWrite(CCoinsMap &mapCoins, const uint256 &hashBlockIn
             it++;
     }
     hashBlock = hashBlockIn;
+    if (nBestCoinHeightIn > nBestCoinHeight)
+        nBestCoinHeight = nBestCoinHeightIn;
+
     return true;
 }
 
 bool CCoinsViewCache::Flush()
 {
     LOCK(cs_utxo);
-    bool fOk = base->BatchWrite(cacheCoins, hashBlock, cachedCoinsUsage);
+    bool fOk = base->BatchWrite(cacheCoins, hashBlock, nBestCoinHeight, cachedCoinsUsage);
     return fOk;
 }
 
 void CCoinsViewCache::Trim(size_t nTrimSize) const
 {
-    uint64_t nTrimmed = 0;
-
     LOCK(cs_utxo);
+
+    uint64_t nTrimmed = 0;
+    uint64_t nTrimmedByHeight = 0;
+    static uint64_t nTrimHeightDelta = nBestCoinHeight * 0.80; // This is where we attempt to do our first trim
+    int64_t nTrimHeight = nBestCoinHeight - nTrimHeightDelta;
+
+    // Begin first Trim loop. This loop will trim coins from cache by the coin height, removing the oldest coins first.
+    // This has been proven to improve sync performance significantly for nodes that can not hold the entire dbcache
+    // in memory.
+    bool fDone = false;
+    int nSmallestDelta = 50; // number of blocks to adjust trim height by
     CCoinsMap::iterator iter = cacheCoins.begin();
+    while (!fDone && DynamicMemoryUsage() > nTrimSize)
+    {
+        LogPrint("coindb", "cacheCoinsUsage at start: %d total dynamic usage: %d trim to size: %d nBestCoinHeight: %d "
+                           "trim height:%d\n",
+            cachedCoinsUsage, DynamicMemoryUsage(), nTrimSize, nBestCoinHeight, nTrimHeight);
+
+        iter = cacheCoins.begin();
+        while (DynamicMemoryUsage() > nTrimSize)
+        {
+            if (iter == cacheCoins.end())
+            {
+                fDone = true;
+                break;
+            }
+
+            if (iter->second.flags == 0 && iter->second.coin.nHeight < nTrimHeight)
+            {
+                cachedCoinsUsage -= iter->second.coin.DynamicMemoryUsage();
+
+                iter = cacheCoins.erase(iter);
+                nTrimmed++;
+                nTrimmedByHeight++;
+            }
+            else
+                iter++;
+        }
+
+        // Gradually increase the nTrimHeight if we didn't trim enought entries.
+        if (fDone && DynamicMemoryUsage() > nTrimSize && nTrimHeightDelta > nSmallestDelta)
+        {
+            if (nTrimHeightDelta <= nSmallestDelta * 100)
```

Comment:
Compiler warning when comparing unsigned and signed integer values
uint64_t nTrimHeightDelta
int  nSmallestDelta

Reply:
thx...fixed.

Reply:
Peter, thanks for the fix. So I did both unit and regression tests

There are failed tests on the regression suite. They may not be related to this pull request
I also ran the regression test on "release" branch, only 1 test "excessive.py  --extended" failed.
Want to confirm if you also got similar failed tests. 
Should I create a separate ticket to track the failed test on regression suite?
Thanks

(1) I ran unit test (coins_tests) and all passed
>test_bitcoin --log_level=all --run_test=coins_tests

(2) Run the regression test suite with qa/pull-tester/rpc-tests.py
Runtime: 214 s
28 test(s) passed / 8 test(s) failed / 36 test(s) executed
4 test(s) disabled / 0 test(s) skipped due to platform

(3) Run all possible tests with qa/pull-tester/rpc-tests.py -extended
Runtime: 16012 s
50 test(s) passed / 6 test(s) failed / 56 test(s) executed
7 test(s) disabled / 0 test(s) skipped due to platform